bsnes/sfc/cpu/timing.cpp

```diff
@@ -200,75 +200,47 @@ auto CPU::dmaEdge() -> void {
   }
 }
 
-//called every 256 clocks; see CPU::step()
+//called every 128 clocks; see CPU::step()
```

Comment:
It seems this is actually called from `CPU::stepOnce()` these days, not `CPU::step()`.

Reply:
Also, your PR comment says:

> The new code runs the auto-joypad polling at 128-cycle granularity instead of 256-cycle granularity, so that the first event can happen ~128 cycles after the start of Vblank.

Why is that desirable, if Jonas Quinn's changes to higan didn't need it? (I'm genuinely confused here)

Reply:
We might as well make the comment CPU::stepOnce, good catch!

anomie's notes were slightly off. So let me back up: 32.5-95.5 is referring to the PPU dot clock. If you multiply by 4 you get the cycle count, 130-382. It's pretty clear that in 256-cycle units, that would need to be 130-384 or you'd have no event when it lands on cycle 384. Jonas Quinn confirmed that with his testing in any case.

I chose to split the cycle range in half and run things every 128 clocks. So that reduces us to 32.5-64.0, or 130 to 256 + 258-384. The reason I chose 128 over 256 was so I could emulate the time it takes to perform the joypad latching. I don't believe you can set the controller latch to 1 and then 0 immediately. It needs time to let the controllers see this happening and then release the latch, and time to let that take effect. Polling the controllers is only one operation so it makes sense that happens once every 256 cycles, which is why I have the (counter&1) test for that.

Think about it this way: we know the polling sequence takes 4224 cycles, and 4224/256 is 16.5. Isn't that a little strange to have half a cycle? But if we consider the latching is actually twice as fast, then it's 2x128 + 16x256. If we run on a 128-clock counter, it becomes 2x128 + 16x2x128, or 34 total states. But now since we don't actually poll on the second half of each 16x2x256 states, we don't need to bother running the event there, so we can stop at 33 instead of 34 and it's the same thing. And by doing that, it turns our polling duration into 4224 cycles, just like we've observed.

Jonas Quinn's patch does the latch to 1 then latch to 0 at the same instant, which is why he can use a 256-cycle counter.

Reply:
That was very well explained, thank you!